x86: build cpukinds from AMD CPUID leaf 0x80000026

bgoglin opened the pull request:
Use coretype to build "AMDEfficiency" and "AMDPerformance" kinds.

Use power efficiency rankings to build kinds with native efficiencies, or just make E=0 and P=1.

bgoglin commented:
@mkuron thanks to your CPUID dump in #690, I was able to verify that this code mostly works:
```
CPU kind #0 efficiency -1 cpuset 0x00ff0ff0
  CoreType = AMD-ECore
CPU kind #1 efficiency -1 cpuset 0x0000f00f
  CoreType = AMD-PCore
```
Does this CPU report EfficiencyRanking? I see 1 in EfficiencyRankingAvailable=1 in the first level but the ranking I read is 0 for all cores (that's why I have "efficiency -1" (unknown) above).

Also, I am not sure I understand why AsymmetricTopology=1 only in the second level, EfficiencyRankingAvailable=1 only in first level (that one is documented), and HeterogeneousCores=1 in first 3 levels but not in 4th.

superm1 commented on hwloc/topology-x86.c:
```diff
@@ -583,9 +585,44 @@ static void read_extended_topo(struct hwloc_x86_backend_data_s *data, struct pro
+              infos->power_efficiency_ranking = (ebx >> 16) & 0xff;
+            }
+
+            infos->hybridcoretype = (ebx >> 28) & 0xf; /* 0 = P, 1 = E */
+            infos->hybridnativemodel = (ebx >> 24) & 0xf; /* 0 = Zen4 */
```
both of these fields are only 4 bits a piece

superm1 commented on hwloc/topology-x86.c:
```diff
@@ -583,9 +585,44 @@ static void read_extended_topo(struct hwloc_x86_backend_data_s *data, struct pro
+              /* "a core with a lower value has intrinsically better power,
+               *  but potentially lower performance potential vs cores with a higher value."
+               */
+              infos->power_efficiency_ranking = (ebx >> 16) & 0xff;
```
this field is only 8 bits

superm1 commented on hwloc/topology-x86.c:
```diff
@@ -583,9 +585,44 @@ static void read_extended_topo(struct hwloc_x86_backend_data_s *data, struct pro
+  printf(" AsymmetricCores = %u\n", (eax & 0x80000000)>>31);
+  printf(" HeteroCoreTyopology  = %u\n", (eax & 0x40000000)>>30);
+  printf(" EffRankingAvailable = %u\n", (eax & 0x20000000)>>29);
+  printf(" CoreType = %u\n", (ebx >> 28) & 0xf);
```
4 bits for this field

superm1 commented on hwloc/topology-x86.c:
```diff
@@ -583,9 +585,44 @@ static void read_extended_topo(struct hwloc_x86_backend_data_s *data, struct pro
+  printf(" HeteroCoreTyopology  = %u\n", (eax & 0x40000000)>>30);
+  printf(" EffRankingAvailable = %u\n", (eax & 0x20000000)>>29);
+  printf(" CoreType = %u\n", (ebx >> 28) & 0xf);
+  printf(" Model = %u\n", (ebx >> 24) & 0xf);
```
4 bits for this field

superm1 commented on hwloc/topology-x86.c:
```diff
@@ -583,9 +585,44 @@ static void read_extended_topo(struct hwloc_x86_backend_data_s *data, struct pro
+  printf(" EffRankingAvailable = %u\n", (eax & 0x20000000)>>29);
+  printf(" CoreType = %u\n", (ebx >> 28) & 0xf);
+  printf(" Model = %u\n", (ebx >> 24) & 0xf);
+  printf(" EffRanking = %u\n", (ebx >> 16) & 0xff);
```
8 bits for this one

bgoglin replied:
Well, 0xf is 4 bits and 0xff is 8bits so all these look OK?

superm1 replied:
mm I guess so

superm1 commented:
> Also, I am not sure I understand why AsymmetricTopology=1 only in the second level

It's a topology indicator only.  IE how they're grouped into larger units.  It's meant to be used in conjunction with bit 30 and 29.  IE if it's "1" then software is supposed to discover more details from 30 and 29.

> EfficiencyRankingAvailable=1 only in first level (that one is documented)

I think APM explains that already.

>  HeterogeneousCores=1 in first 3 levels but not in 4th

It's a single socket system.

bgoglin commented:
@superm1 Is the "NativeModelId" CPUID info supposed to be the same for all cores on hybrid CPUs?
The doc says "This value is used by software to further differentiate implementation specific software visible properties between the cores". I am not sure if that applies to cores of different chips, or different cores of the same chip.

I couldn't find the PPR for Strix Point. In the 19h family PPR, I see "Zen4" as the only NativeModelID possible value. Could we see some hybrid CPUs return something like "Zen5" and "Zen5c" for different cores? Or will they all return "Zen5" while CoreType is the only field distinguishing between Ecore and Pcore?

By the way, in case you missed the question above:
Does this CPU report EfficiencyRanking? I see 1 in EfficiencyRankingAvailable=1 in the first level but the ranking I read is 0 for all cores (that's why I have "efficiency -1" (unknown) above).

superm1 commented:
Yeah the publication of the docs sometimes lags a bit from the launch, sorry about that.  I double checked though and  `0h` means `Zen5` for `NativeModelId` on Strix.  There is no differentiation of Zen5 vs Zen5c at this field.  You would tell that from `CoreType` when called with the cpuid is called with level `1h (Core)`

> Does this CPU report EfficiencyRanking? I see 1 in EfficiencyRankingAvailable=1 in the first level but the ranking I read is 0 for all cores (that's why I have "efficiency -1" (unknown) above).

I guess not; but I'm not sure why.  There are definitely rankings utilized from the CPPC MSRs.

bgoglin committed:
x86: abstract out the function to build Intel cpukinds

Signed-off-by: Brice Goglin <Brice.Goglin@inria.fr>

bgoglin committed:
x86: use 0x80000026 CPUID leaf to detect AMD Performance and Efficiency cores

The leaf tells whether cores have different types (we use that),
different number of threads (we don't), and provide efficiency
ranking (although that's not the case on some of the first generation CPU).

For now, we only support 2 kinds (E and P), even if rankings are not
provided.

If AMD processors with more kinds appear in the future, hopefully
the ranking will be correct and we'll be able to distinguish
Ecores from LP-Ecores easily.

Signed-off-by: Brice Goglin <Brice.Goglin@inria.fr>

bgoglin committed:
tests/x86: add an AMD hybrid CPU

Signed-off-by: Brice Goglin <Brice.Goglin@inria.fr>

bgoglin committed:
showmsg: add x86 flag

Signed-off-by: Brice Goglin <Brice.Goglin@inria.fr>

bgoglin committed:
x86: warn if we find unexpected Intel core types

Signed-off-by: Brice Goglin <Brice.Goglin@inria.fr>
